src/sparse_merkle_tree/smt.rs

```diff
@@ -842,38 +842,78 @@ impl SparseMerkleTree {
         Ok((query_with_proofs, ancestor_hashes))
     }
 
-    fn verify_query_keys(proof: &Proof, query_keys: &[Vec<u8>], key_length: KeyLength) -> bool {
-        let mut queries: HashMap<&[u8], QueryProof> = HashMap::new();
+    /// verify_and_prepare_proof_map checks all verifications of query_keys based on the verify function in the [LIP-0039](https://github.com/LiskHQ/lips/blob/main/proposals/lip-0039.md#proof-construction).
+    /// Also it returns a tuple, the first parameter is the result of the query_keys verification and the second one is a map
+    fn verify_and_prepare_proof_map(
+        proof: &Proof,
+        query_keys: &[Vec<u8>],
+        key_length: KeyLength,
+    ) -> Result<HashMap<Vec<bool>, QueryProofWithProof>, SMTError> {
+        let mut queries_with_proof: HashMap<Vec<bool>, QueryProofWithProof> = HashMap::new();
+
+        if query_keys.len() != proof.queries.len() {
+            return Err(SMTError::InvalidInput(String::from(
+                "Mismatched length of keys and the queries of the proof",
+            )));
+        }
+        let mut queries: HashMap<Vec<bool>, QueryProof> = HashMap::new();
         for (i, key) in query_keys.iter().enumerate() {
-            if key.len() != key_length.into() {
-                return false;
+            // Check if all the query keys have the same length.
+            if proof.queries[i].key().len() != key_length.into() || key.len() != key_length.into()
+            {
+                return Err(SMTError::InvalidInput(String::from(
+                    "The length of the key is invalid",
+                )));
             }
+
             let query = &proof.queries[i];
-            let duplicate_query = queries.get(query.key());
-            if let Some(q) = duplicate_query {
-                if !utils::is_bytes_equal(&q.bitmap, &query.bitmap)
-                    || !utils::is_bytes_equal(q.value(), query.value())
+            let binary_bitmap = utils::strip_left_false(&utils::bytes_to_bools(&query.bitmap));
+            let query_key_binary = utils::bytes_to_bools(query.key());
+            let binary_path = if binary_bitmap.len() > query_key_binary.len() {
```

Comment:
I think this check should be done only if the query key and the relative proof query key are unequal, just like indicated in the LIP